be/src/olap/storage_engine.cpp

```diff
@@ -1002,4 +956,60 @@ OLAPStatus StorageEngine::load_header(
     return res;
 }
 
+
+OLAPStatus StorageEngine::execute_task(EngineTask* task) {
+    // 1. add wlock to related tablets
+    // 2. do prepare work
+    // 3. release wlock
+    {
+        vector<TabletInfo> tablet_infos;
+        task->get_related_tablets(&tablet_infos);
+        sort(tablet_infos.begin(), tablet_infos.end());
+        for (TabletInfo& tablet_info : tablet_infos) {
+            TabletSharedPtr tablet = TabletManager::instance()->get_tablet(
+                tablet_info.tablet_id, tablet_info.schema_hash, false);
+            tablet->obtain_header_wrlock();
+        }
+        // add write lock to all related tablets
+        OLAPStatus prepare_status = task->prepare();
+        for (TabletInfo& tablet_info : tablet_infos) {
+            TabletSharedPtr tablet = TabletManager::instance()->get_tablet(
+                tablet_info.tablet_id, tablet_info.schema_hash, false);
+            tablet->release_header_lock();
+        }
+        if (prepare_status != OLAP_SUCCESS) {
+            return prepare_status;
+        }
+    }
+
+    // do execute work without lock
+    OLAPStatus exec_status = task->execute();
+    if (exec_status != OLAP_SUCCESS) {
+        return exec_status;
+    }
+    
+    // 1. add wlock to related tablets
+    // 2. do finish work
+    // 3. release wlock
+    {
+        vector<TabletInfo> tablet_infos;
+        // related tablets may be changed after execute task, so that get them here again
+        task->get_related_tablets(&tablet_infos);
+        sort(tablet_infos.begin(), tablet_infos.end());
+        for (TabletInfo& tablet_info : tablet_infos) {
+            TabletSharedPtr tablet = TabletManager::instance()->get_tablet(
+                tablet_info.tablet_id, tablet_info.schema_hash, false);
+            tablet->obtain_header_wrlock();
+        }
+        // add write lock to all related tablets
+        OLAPStatus fin_status = task->finish();
+        for (TabletInfo& tablet_info : tablet_infos) {
+            TabletSharedPtr tablet = TabletManager::instance()->get_tablet(
+                tablet_info.tablet_id, tablet_info.schema_hash, false);
+            tablet->release_header_lock();
+        }
```

Comment:
I think these lines should be in a function because they are called at least twice